fix(gist): add support for editing & viewing large files

luxass opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->
resolves #11739

luxass committed:
feat(gist): add `RawURL` and `Truncated` fields to `GistFile`

* Added `RawURL` to store the raw URL of the gist file.
* Added `Truncated` to indicate if the content is truncated.

luxass committed:
feat(gist): add `GetRawGistFile` function to retrieve raw gist content

* Implements a new function to fetch the raw content of a gist using its `rawURL`.
* Handles HTTP requests and responses, including error management for non-200 status codes.

luxass committed:
feat(gist): retrieve full content for truncated gist files

* Added logic to fetch the full content of a gist file if it is marked as truncated.
* Utilizes the `GetRawGistFile` function to obtain the complete content from the provided `RawURL`.

luxass committed:
feat(gist): retrieve full content for truncated gist files

* Added logic to fetch full content for truncated files when updating a gist.
* Utilizes `shared.GetRawGistFile` to retrieve the complete content based on the `RawURL`.

luxass committed:
test(gist): add tests for `GetRawGistFile` function

- Implemented unit tests for various scenarios in `GetRawGistFile`.
- Covered cases including successful requests, error handling, and different content types.
- Ensured proper verification of HTTP responses and error messages.

luxass committed:
test(gist): add tests for editing truncated gist files

* Implement tests for editing gists that contain truncated files.
* Ensure that both single and multiple truncated files are handled correctly.
* Mock responses for retrieving full content from raw URLs for truncated files.

luxass committed:
test(gist): add tests for handling truncated files in view command

* Added test cases for viewing truncated files with and without the raw flag.
* Included scenarios for multiple files with truncation behavior.
* Ensured correct output is returned for truncated files when accessed via raw URLs.

luxass committed:
Merge branch 'trunk' into gist-edit-large-file

BagToad reviewed (commented):
Hey @luxass 👋 Thanks for your patience!

I've been reviewing this with @babakks and it's pretty good, but we're concerned about fetching the full gist content of each file attached to a gist even when the user hasn't selected to edit it.

We think it would be better to fetch only the raw content of a particular gist when the user has selected to edit a gist where `truncated` is `true`. If the user doesn't select a gist where `truncated` is true, we don't need to fetch the full raw file content.

As we investigated and reviewed further, I realized that we're also sending the full list of files to [the API](https://docs.github.com/en/rest/gists/gists?apiVersion=2022-11-28#update-a-gist) anytime we post an update, even though some are unchanged 🤔 I believe this to be part of the underlying cause of the data loss reported in #11739 since any file that is truncated will be updated with the truncated contents instead of the full contents:

> Files from the previous version of the gist that aren't explicitly changed during an edit are unchanged.

I believe this documentation point means we could omit gist files from the upload payload that the user doesn't change. Doing so in conjunction with only fetching raw gist content when needed by the user seems like the ideal experience.

I think we need to do some testing to ensure I'm interpreting the API documentation correctly, and then if so, change our approach to:

- only fetch raw gist content when needed
- only include files that have changed in the gist update request

Let me know if that makes sense and if you're interested in helping out with this still 🙏

luxass committed:
Merge branch 'trunk' into gist-edit-large-file

luxass commented:
Hi @BagToad

That makes total sense 👍🏻 I will continue work on this PR.

luxass committed:
fix(gist): fetch full content for truncated files during editing

This change makes it only fetch the raw content, when we wanna edit the specific file.

luxass committed:
fix(gist): only include files that have changed in request

There was some issues when sending the full list of files, some of them would be truncated.

luxass committed:
fix(gist): prevent fetching full content for already edited files

luxass committed:
test(gist): fix failing tests

luxass commented:
The only thing I'm missing is some more tests - otherwise i think it is ready to go. 🙂

I've verified that the behavior

> Files from the previous version of the gist that aren't explicitly updated during an edit remain unchanged

is correct and working 👍🏻

BagToad committed:
Merge branch 'trunk' into gist-edit-large-file

BagToad commented:
Thank you very much @luxass! ✨ And great work!

> The only thing I'm missing is some more tests - otherwise i think it is ready to go. 🙂

I took a stab at filling in the gaps on the tests, feel free to commit if you think they look good:

<details><summary>Show Diff</summary>
<p>

```diff
diff --git a/pkg/cmd/gist/edit/edit_test.go b/pkg/cmd/gist/edit/edit_test.go
index 296ae3261..ac1555f5c 100644
--- a/pkg/cmd/gist/edit/edit_test.go
+++ b/pkg/cmd/gist/edit/edit_test.go
@@ -237,6 +237,33 @@ func Test_editRun(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "single file edit flag sends only edited file",
+			opts: &EditOptions{
+				Selector:     "1234",
+				EditFilename: "unix.md",
+			},
+			mockGist: &shared.Gist{
+				ID: "1234",
+				Files: map[string]*shared.GistFile{
+					"cicada.txt": {Filename: "cicada.txt", Content: "bwhiizzzbwhuiiizzzz", Type: "text/plain"},
+					"unix.md":    {Filename: "unix.md", Content: "meow", Type: "text/markdown"},
+				},
+				Owner: &shared.GistOwner{Login: "octocat"},
+			},
+			httpStubs: func(reg *httpmock.Registry) {
+				reg.Register(httpmock.REST("POST", "gists/1234"), httpmock.StatusStringResponse(201, "{}"))
+			},
+			wantLastRequestParameters: map[string]interface{}{
+				"description": "",
+				"files": map[string]interface{}{
+					"unix.md": map[string]interface{}{
+						"content":  "new file content",
+						"filename": "unix.md",
+					},
+				},
+			},
+		},
 		{
 			name:  "multiple files, cancel, with TTY",
 			isTTY: true,
@@ -645,6 +672,48 @@ func Test_editRun(t *testing.T) {
 				},
 			},
 		},
+		{
+			name:  "interactive truncated multi-file gist fetches only selected file raw content the first time",
+			isTTY: true,
+			opts:  &EditOptions{Selector: "1234"},
+			prompterStubs: func(pm *prompter.MockPrompter) {
+				pm.RegisterSelect("Edit which file?", []string{"also-truncated.txt", "large.txt"}, func(_, _ string, opts []string) (int, error) {
+					return prompter.IndexFor(opts, "large.txt")
+				})
+				pm.RegisterSelect("What next?", editNextOptions, func(_, _ string, opts []string) (int, error) {
+					return prompter.IndexFor(opts, "Edit another file")
+				})
+				// Editing large.txt twice to ensure that fetch for the raw URL happens only once
+				pm.RegisterSelect("Edit which file?", []string{"also-truncated.txt", "large.txt"}, func(_, _ string, opts []string) (int, error) {
+					return prompter.IndexFor(opts, "large.txt")
+				})
+				pm.RegisterSelect("What next?", editNextOptions, func(_, _ string, opts []string) (int, error) {
+					return prompter.IndexFor(opts, "Submit")
+				})
+			},
+			mockGist: &shared.Gist{
+				ID: "1234",
+				Files: map[string]*shared.GistFile{
+					"large.txt":          {Filename: "large.txt", Content: "This is truncated content...", Type: "text/plain", Truncated: true, RawURL: "https://gist.githubusercontent.com/user/1234/raw/large.txt"},
+					"also-truncated.txt": {Filename: "also-truncated.txt", Content: "stuff...", Type: "text/plain", Truncated: true, RawURL: "https://gist.githubusercontent.com/user/1234/raw/also-truncated.txt"},
+				},
+				Owner: &shared.GistOwner{Login: "octocat"},
+			},
+			httpStubs: func(reg *httpmock.Registry) {
+				reg.Register(httpmock.REST("POST", "gists/1234"), httpmock.StatusStringResponse(201, "{}"))
+				// Explicity exclude also-truncated.txt raw URL to ensure it is not fetched since we did not select it.
+				reg.Exclude(t, httpmock.REST("GET", "user/1234/raw/also-truncated.txt"))
+			},
+			wantLastRequestParameters: map[string]interface{}{
+				"description": "",
+				"files": map[string]interface{}{
+					"large.txt": map[string]interface{}{
+						"content":  "new file content",
+						"filename": "large.txt",
+					},
+				},
+			},
+		},
 	}
 
 	for _, tt := range tests {
```


</p>
</details>

luxass committed:
test(gist/edit): add tests for single file edit and interactive multi-file selection

Co-authored-by: Kynan Ware <47394200+BagToad@users.noreply.github.com>

luxass commented:
Thanks a lot for adding those tests - I've pushed them 👍🏻 


They were very close to what I had in mind, but I couldn't quite find a nice way to assert that the raw URL was only hit the first time. Your version nailed that part, so thanks 🙌🏻

BagToad reviewed (approved):
Alright I think this looks great! Thanks much for the collaboration on this one! 🍻
